<a href="https://colab.research.google.com/github/JulioRan09/Porgramacion-Predictiva/blob/main/Actividad_Calificada_07_Datos_Temporales_276572.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Actividad calificada: Datos temporales**
### Limpieza y transformación de fechas, horas y duraciones en un dataset real

---

| | |
|---|---|
| **Materia** | Programación para Analítica Descriptiva y Predictiva |
| **Programa** | Maestría en Inteligencia Artificial y Analítica de Datos (MIAAD) — UACJ |
| **Unidad** | 02 Análisis Descriptivo de los Datos |
| **Tema** | Limpieza y transformación de datos: datos temporales |
| **Modalidad** | Individual |
| **Valor** | 100 puntos |
| **Dataset** | Viajes de taxi en Nueva York, marzo de 2019 (NYC TLC, muestra publicada en un repositorio público de GitHub) |
| **Fecha de entrega** | 26 septiembre |

---

## Datos del estudiante

| | |
|---|---|
| **Nombre completo** | Julio Rangel García |
| **Matrícula** | 276572 |

## Instrucciones

1. Trabaja en Google Colab. Guarda una copia en tu Drive (**Archivo → Guardar una copia en Drive**) antes de empezar.
2. Completa tus datos en la tabla anterior.
3. Resuelve las actividades **en orden**: algunas usan variables creadas en actividades anteriores.
4. Escribe tu código en las celdas marcadas con `# Escribe tu código aquí`. Puedes agregar celdas si lo necesitas.
5. **Comenta tu código**: cada bloque debe explicar qué hace y por qué. Los comentarios forman parte de la calificación de cada actividad.
6. Responde las preguntas en las celdas **Respuesta**, con tus propias palabras y apoyándote en tus resultados (cita cifras).
7. Usa los métodos vistos en la sesión: `pd.to_datetime()`, el accesor `.dt`, `pd.Timedelta`, `.between()`, `tz_localize()`, `groupby()`, `sort_values()`, `.where()`, máscaras booleanas, etc.
8. Antes de entregar, ejecuta todo el notebook (**Entorno de ejecución → Reiniciar y ejecutar todo**) y verifica que no haya errores.
9. **Entrega en Moodle** un archivo de texto `Actividad_Temporales_<Matricula>.txt` que contenga el **enlace** a tu notebook (Google Colab o GitHub), con las salidas visibles.
   - Colab: **Compartir → Acceso general → Cualquier persona con el enlace (Lector)**.
   - GitHub: el repositorio debe ser **público**.
   - Opcional: puedes adjuntar también el archivo `.ipynb`.

## Contexto

El dataset contiene 6,433 viajes de taxis amarillos y verdes de la ciudad de Nueva York registrados en **marzo de 2019**. Cada fila es un viaje. Las fechas y horas están en **hora local de Nueva York**.

| Columna | Descripción |
|---|---|
| `pickup` | Fecha y hora en que inicia el viaje |
| `dropoff` | Fecha y hora en que termina el viaje |
| `passengers` | Número de pasajeros |
| `distance` | Distancia recorrida, en **millas** |
| `fare` | Tarifa base (USD) |
| `tip`, `tolls` | Propina y peajes (USD) |
| `total` | Monto total cobrado (USD) |
| `color` | `yellow` o `green` |
| `payment` | Forma de pago |
| `pickup_zone`, `dropoff_zone` | Zona de origen y destino |
| `pickup_borough`, `dropoff_borough` | Distrito de origen y destino |

> **Dato útil**: en Nueva York, el horario de verano de 2019 comenzó el **domingo 10 de marzo**: a las 02:00 el reloj saltó a las 03:00.

## Rúbrica

| Actividad | Puntos | Excelente | Bueno | Insuficiente | No entregado / Incorrecto |
|---|---|---|---|---|---|
| **1. Conversión y validación** | 20 | 18–20: conversión con formato explícito y verificación de `NaT`; viajes fuera de periodo detectados; análisis del cambio de horario correcto; código comentado; conclusión con cifras. | 14–17: resultados correctos con omisiones menores en comentarios o conclusión. | 1–13: conversión sin verificar o validaciones incompletas; conclusión sin sustento. | 0: no se entregó o el código no produce resultados. |
| **2. Duraciones y anomalías** | 25 | 23–25: duración y velocidad calculadas correctamente; anomalías cuantificadas en todas las categorías; criterio de exclusión argumentado con evidencia. | 18–22: cálculos correctos; criterio parcial o con poca evidencia. | 1–17: cálculos con errores o anomalías no cuantificadas. | 0: no se entregó o es incorrecto. |
| **3. Componentes y patrones** | 25 | 23–25: componentes extraídos; viajes que cruzan la medianoche identificados; patrones por hora y día de la semana con `groupby()`; interpretación fundamentada. | 18–22: componentes y agrupaciones correctos; interpretación incompleta. | 1–17: agrupaciones con errores o sin interpretación. | 0: no se entregó o es incorrecto. |
| **4. Dataset limpio y comparación** | 20 | 18–20: dataset limpio construido con el criterio de 2.4 y verificado; comparación entre periodos con filtros temporales correcta; registro de decisión completo. | 14–17: dataset limpio correcto; comparación o registro incompletos. | 1–13: criterio aplicado de forma inconsistente o sin verificación. | 0: no se entregó o es incorrecto. |
| **Reflexión final** | 10 | 9–10: respuestas fundamentadas en los resultados propios. | 7–8: respuestas correctas pero generales. | 1–6: respuestas vagas o sin relación con los resultados. | 0: no se entregó. |
| **Total** | **100** | | | | |

**Notas**:

- Las respuestas escritas de cada actividad (1.4, 2.4, 3.4, 4.3) y los comentarios del código cuentan para la calificación **de esa actividad**, no solo para la reflexión final.
- Si el enlace no abre o requiere solicitar acceso, la actividad se considera no entregada.
- Entrega tardía: se descuenta **5 % por día natural** de retraso, hasta 7 días; después de ese plazo la calificación es 0. El descuento se aplica sobre la calificación obtenida.

## Carga de datos

Ejecuta la siguiente celda. No requiere modificaciones.

In [1]:
import pandas as pd

# Carga del dataset de taxis de Nueva York desde un repositorio público de GitHub.
# El archivo se lee con pandas; no se requiere ninguna otra biblioteca.
url = 'https://raw.githubusercontent.com/mwaskom/seaborn-data/master/taxis.csv'
taxis = pd.read_csv(url)

print('Dimensiones:', taxis.shape)
taxis.head()

Dimensiones: (6433, 14)


,pickup,dropoff,passengers,distance,fare,tip,tolls,total,color,payment,pickup_zone,dropoff_zone,pickup_borough,dropoff_borough
0,2019-03-23 20:21:09,2019-03-23 20:27:24,1,1.60,7.0,2.15,0.0,12.95,yellow,credit card,Lenox Hill West,UN/Turtle Bay South,Manhattan,Manhattan
1,2019-03-04 16:11:55,2019-03-04 16:19:00,1,0.79,5.0,0.00,0.0,9.30,yellow,cash,Upper West Side South,Upper West Side South,Manhattan,Manhattan
2,2019-03-27 17:53:01,2019-03-27 18:00:25,1,1.37,7.5,2.36,0.0,14.16,yellow,credit card,Alphabet City,West Village,Manhattan,Manhattan
3,2019-03-10 01:23:59,2019-03-10 01:49:51,1,7.70,27.0,6.15,0.0,36.95,yellow,credit card,Hudson Sq,Yorkville West,Manhattan,Manhattan
4,2019-03-30 13:27:42,2019-03-30 13:37:14,3,2.16,9.0,1.10,0.0,13.40,yellow,credit card,Midtown East,Yorkville West,Manhattan,Manhattan


---
## Actividad 1. Conversión y validación (20 puntos)

### 1.1 Conversión (5 pts)

1. Muestra el tipo de dato de `pickup` y `dropoff`.
2. Convierte ambas columnas a tipo fecha con `pd.to_datetime()`, indicando el **formato explícito** que corresponde a los valores.
3. Verifica que la conversión no generó `NaT`: compara el número de faltantes antes y después.

In [2]:
# Escribe tu código aquí
#1.1.1 Utilizamos dtype sobre las columnas solicitadas

print(taxis['pickup'].dtype)
print(taxis['dropoff'].dtype)

object
object


In [3]:
#1.1.2 utilizamos el datetime de pandas para convertir y le especificamos el formato

taxis['pickup'] = pd.to_datetime(taxis['pickup'], format='%Y-%m-%d %H:%M:%S')
taxis['dropoff'] = pd.to_datetime(taxis['dropoff'], format='%Y-%m-%d %H:%M:%S')
print(taxis['pickup'].dtype)
print(taxis['dropoff'].dtype)

datetime64[ns]
datetime64[ns]


In [4]:
#1.1.3 Confirmamos que no hay nigún nulo en las columnas

print(taxis['pickup'].isna().sum())
print(taxis['dropoff'].isna().sum())

0
0


### 1.2 Periodo de los datos (5 pts)

El dataset debería contener solo viajes de marzo de 2019.

1. Muestra la fecha mínima y máxima de `pickup` y de `dropoff`.
2. Cuenta y muestra los viajes cuyo `pickup` **no** está en marzo de 2019.
3. ¿Hay viajes con `pickup` en marzo y `dropoff` fuera de marzo? Cuéntalos.

In [5]:
# Escribe tu código aquí
#1.2.1 Utilizamos min y max para encontrar las fechas que nos solicita la actividad

min_pickup = taxis['pickup'].min()
max_pickup = taxis['pickup'].max()
min_dropoff = taxis['dropoff'].min()
max_dropoff = taxis['dropoff'].max()

#Descubrimos que hay viajes de febrero y mayo, no sólo de marzo
print(f'La fecha mínima de Pick up es {min_pickup}')
print(f'La fecha máxima de Pick up es {max_pickup}')
print(f'La fecha mínima de Drop off es {min_dropoff}')
print(f'La fechá máxima de Drop off es {max_dropoff}')

La fecha mínima de Pick up es 2019-02-28 23:29:03
La fecha máxima de Pick up es 2019-03-31 23:43:45
La fecha mínima de Drop off es 2019-02-28 23:32:35
La fechá máxima de Drop off es 2019-04-01 00:13:58


In [6]:
#1.2.2 Utilizamos una condición para encontrar el dato fuera de Marzo
pick_feb = taxis.copy()
pick_feb = taxis[taxis['pickup']< '2019-03-01']

print('La cantidad de datos con pickup fuera de marzo es:', pick_feb.shape[0], 'los datos se muestran a continuación:\n')
pick_feb.head()


La cantidad de datos con pickup fuera de marzo es: 1 los datos se muestran a continuación:



,pickup,dropoff,passengers,distance,fare,tip,tolls,total,color,payment,pickup_zone,dropoff_zone,pickup_borough,dropoff_borough
6203,2019-02-28 23:29:03,2019-02-28 23:32:35,1,0.9,5.0,0.0,0.0,6.3,green,cash,Old Astoria,Long Island City/Queens Plaza,Queens,Queens


In [7]:
#1.2.3 Encontramos que hay un viaje que se tomó el 31 de marzo y se terminó el 1 de mayo en la madrugada
pick_march = taxis.copy()
pick_march = taxis[(taxis['pickup'] >= '2019-03-01') & (taxis['dropoff'] >= '2019-04-01')]
pick_march

,pickup,dropoff,passengers,distance,fare,tip,tolls,total,color,payment,pickup_zone,dropoff_zone,pickup_borough,dropoff_borough
591,2019-03-31 23:43:45,2019-04-01 00:13:58,5,12.25,37.0,0.0,0.0,40.8,yellow,cash,LaGuardia Airport,Williamsburg (South Side),Queens,Brooklyn


### 1.3 Cambio de horario (5 pts)

El 10 de marzo de 2019, en Nueva York, las horas entre 02:00 y 02:59 **no existieron**.

1. Filtra los viajes con `pickup` el 10 de marzo y cuenta cuántos hay en cada hora (`.dt.hour` y `groupby()` o `value_counts()`). ¿Hay viajes a las 2:00?
2. Localiza `pickup` en la zona `'America/New_York'` con `tz_localize()` usando `nonexistent='NaT'` y cuenta cuántos valores quedan como `NaT`.

In [8]:
# Escribe tu código aquí
#1.3.1 Encontramos que no hay viajes realizados a las 2:00, la columna pickup salta el valor "2"
viajes_march_10 = taxis[(taxis['pickup'] >= '2019-03-10 00:00:00') & (taxis['pickup'] < '2019-03-11 00:00:00')]
viajes_march_10.groupby(viajes_march_10['pickup'].dt.hour).count()

,pickup,dropoff,passengers,distance,fare,tip,tolls,total,color,payment,pickup_zone,dropoff_zone,pickup_borough,dropoff_borough
pickup,,,,,,,,,,,,,,
0,11,11,11,11,11,11,11,11,11,11,11,10,11,10
1,7,7,7,7,7,7,7,7,7,7,7,7,7,7
3,6,6,6,6,6,6,6,6,6,6,6,6,6,6
4,7,7,7,7,7,7,7,7,7,7,7,7,7,7
5,1,1,1,1,1,1,1,1,1,1,1,1,1,1
6,2,2,2,2,2,2,2,2,2,2,2,2,2,2
7,5,5,5,5,5,5,5,5,5,5,5,5,5,5
8,7,7,7,7,7,7,7,7,7,7,7,7,7,7
9,9,9,9,9,9,9,9,9,9,9,8,8,8,8


In [9]:
#1.3.2 No se generó ningún dato NaT porque todos los registros tiene correctamente asignada la fecha.
#No hay viajes entre las 2:00 y 2:59 que tuvieran faltantes.
faltantes = viajes_march_10['pickup'].dt.tz_localize('America/New_York', nonexistent='NaT')
print(f'La cantidad de datos faltantes, y que se cambiaron por la leyenda "NaT" es: {faltantes.isna().sum()}')

La cantidad de datos faltantes, y que se cambiaron por la leyenda "NaT" es: 0


### 1.4 Conclusión (5 pts)

¿Qué problemas temporales encontraste en esta actividad? ¿Qué indica el resultado de 1.3 sobre la zona horaria en que están registrados los datos? Responde con cifras.

**Respuesta 1.4:**

_El dataframe presentaba las columnas de pickup y dropoff como objetos, hemos arreglado esto cambiándolas a un formato datetime. Por otro lado, los resultados del ejercicio 1.3 son coherentes entre sí, no se encontraron registros entre las 2:00 y 2:59 en la actividad 1.3.1, y esto se confirmó en 1.3.2 donde al contar cuántos NaT se generaron encontramos que efectivamente, no se generó ningún NaT._

---
## Actividad 2. Duraciones y anomalías (25 puntos)

### 2.1 Duración de cada viaje (6 pts)

1. Crea la columna `duracion` = `dropoff` − `pickup`.
2. Crea `duracion_min` con la duración en **minutos** (usa `.dt.total_seconds()`).
3. Muestra el resumen estadístico de `duracion_min` con `describe()`.

In [10]:
# Escribe tu código aquí
taxis['duracion'] = taxis['dropoff'] - taxis['pickup']
taxis['duracion_min'] = taxis['duracion'].dt.total_seconds()/60
taxis.head()

,pickup,dropoff,passengers,distance,fare,tip,tolls,total,color,payment,pickup_zone,dropoff_zone,pickup_borough,dropoff_borough,duracion,duracion_min
0,2019-03-23 20:21:09,2019-03-23 20:27:24,1,1.60,7.0,2.15,0.0,12.95,yellow,credit card,Lenox Hill West,UN/Turtle Bay South,Manhattan,Manhattan,0 days 00:06:15,6.250000
1,2019-03-04 16:11:55,2019-03-04 16:19:00,1,0.79,5.0,0.00,0.0,9.30,yellow,cash,Upper West Side South,Upper West Side South,Manhattan,Manhattan,0 days 00:07:05,7.083333
2,2019-03-27 17:53:01,2019-03-27 18:00:25,1,1.37,7.5,2.36,0.0,14.16,yellow,credit card,Alphabet City,West Village,Manhattan,Manhattan,0 days 00:07:24,7.400000
3,2019-03-10 01:23:59,2019-03-10 01:49:51,1,7.70,27.0,6.15,0.0,36.95,yellow,credit card,Hudson Sq,Yorkville West,Manhattan,Manhattan,0 days 00:25:52,25.866667
4,2019-03-30 13:27:42,2019-03-30 13:37:14,3,2.16,9.0,1.10,0.0,13.40,yellow,credit card,Midtown East,Yorkville West,Manhattan,Manhattan,0 days 00:09:32,9.533333


In [11]:
taxis['duracion_min'].describe()

,duracion_min
count,6433.000000
mean,14.349617
std,11.643892
min,0.000000
25%,6.500000
50%,10.900000
75%,18.516667
max,107.666667


El comando describe() nos muestra que el promedio de duración de los viajes (en minutos) es de 14.35 minutos, otro dato interesante es que el viaje que más duró fue de 106.67 minutos, casi 2 horas.

### 2.2 Duraciones anómalas (6 pts)

Cuenta los viajes con:

- Duración **negativa** (termina antes de iniciar).
- Duración **cero**.
- Duración **menor a 1 minuto** (incluye los de duración cero).
- Duración **mayor a 3 horas**.

Muestra los viajes de duración cero con las columnas `pickup`, `dropoff`, `distance`, `fare` y `total`.

In [12]:
# Escribe tu código aquí
anomalias = taxis[(taxis['duracion_min'] < 0) | (taxis['duracion_min'] == 0) | (taxis['duracion_min'] < 1) | (taxis['duracion_min'] > 180)]
columnas_analisis = ['pickup', 'dropoff', 'distance', 'fare', 'total', 'duracion_min']

anomalias = anomalias[columnas_analisis].sort_values(by='duracion_min', ascending=False)
print(anomalias.shape)

anomalias.head(10)

(72, 6)


,pickup,dropoff,distance,fare,total,duracion_min
5937,2019-03-01 12:04:25,2019-03-01 12:05:24,0.15,7.0,8.40,0.983333
2811,2019-03-11 17:26:48,2019-03-11 17:27:42,0.20,3.0,8.30,0.900000
624,2019-03-05 09:59:42,2019-03-05 10:00:34,0.31,3.0,7.30,0.866667
5546,2019-03-05 09:43:23,2019-03-05 09:44:15,0.00,2.5,3.30,0.866667
1219,2019-03-24 03:33:20,2019-03-24 03:34:10,0.31,3.0,8.16,0.833333
644,2019-03-30 21:50:51,2019-03-30 21:51:40,0.30,3.0,8.80,0.816667
1851,2019-03-02 19:20:18,2019-03-02 19:21:06,0.10,2.5,5.80,0.800000
5750,2019-03-07 09:17:31,2019-03-07 09:18:18,0.10,2.5,3.30,0.783333
5925,2019-03-16 08:46:29,2019-03-16 08:47:13,0.16,10.0,12.96,0.733333
5460,2019-03-30 03:38:08,2019-03-30 03:38:50,0.07,2.5,3.80,0.700000


### 2.3 Velocidad promedio (6 pts)

La velocidad promedio de un viaje, en millas por hora, es `distance / (duracion_min / 60)`.

1. Crea la columna `velocidad_mph` **solo para viajes con duración mayor que cero**; en los demás debe quedar como faltante (`NaN`). Sugerencia: calcula la velocidad para todos los viajes y conserva solo los valores válidos con `.where(condición)`, que deja `NaN` donde la condición es `False`.
2. Cuenta los viajes con velocidad mayor a 60 mph (velocidad poco plausible en tráfico urbano).
3. Muestra los 5 viajes con mayor velocidad, ordenados de mayor a menor, con `pickup`, `dropoff`, `duracion_min`, `distance`, `velocidad_mph` y `fare`.

In [13]:
# Escribe tu código aquí
taxis['velocidad_mph'] = taxis['distance'] / (taxis['duracion_min'] / 60)
taxis['velocidad_mph'] = taxis['velocidad_mph'].where(taxis['duracion_min'] > 0)
taxis.head()

,pickup,dropoff,passengers,distance,fare,tip,tolls,total,color,payment,pickup_zone,dropoff_zone,pickup_borough,dropoff_borough,duracion,duracion_min,velocidad_mph
0,2019-03-23 20:21:09,2019-03-23 20:27:24,1,1.60,7.0,2.15,0.0,12.95,yellow,credit card,Lenox Hill West,UN/Turtle Bay South,Manhattan,Manhattan,0 days 00:06:15,6.250000,15.360000
1,2019-03-04 16:11:55,2019-03-04 16:19:00,1,0.79,5.0,0.00,0.0,9.30,yellow,cash,Upper West Side South,Upper West Side South,Manhattan,Manhattan,0 days 00:07:05,7.083333,6.691765
2,2019-03-27 17:53:01,2019-03-27 18:00:25,1,1.37,7.5,2.36,0.0,14.16,yellow,credit card,Alphabet City,West Village,Manhattan,Manhattan,0 days 00:07:24,7.400000,11.108108
3,2019-03-10 01:23:59,2019-03-10 01:49:51,1,7.70,27.0,6.15,0.0,36.95,yellow,credit card,Hudson Sq,Yorkville West,Manhattan,Manhattan,0 days 00:25:52,25.866667,17.860825
4,2019-03-30 13:27:42,2019-03-30 13:37:14,3,2.16,9.0,1.10,0.0,13.40,yellow,credit card,Midtown East,Yorkville West,Manhattan,Manhattan,0 days 00:09:32,9.533333,13.594406


In [14]:
#2.3.2
print('La cantidad de viajes con velocidad mayor a 60 mph es:', taxis[taxis['velocidad_mph'] > 60].shape[0])

La cantidad de viajes con velocidad mayor a 60 mph es: 11


In [15]:
#2.3.3
columnas_analisis = ['pickup', 'dropoff', 'duracion_min', 'distance', 'velocidad_mph', 'fare']
taxis.sort_values(by='velocidad_mph', ascending=False)[columnas_analisis].head(5)

,pickup,dropoff,duracion_min,distance,velocidad_mph,fare
770,2019-03-02 03:16:59,2019-03-02 03:17:06,0.116667,9.40,4834.285714,2.5
6089,2019-03-17 23:16:34,2019-03-17 23:16:52,0.300000,10.10,2020.000000,12.0
491,2019-03-07 07:11:33,2019-03-07 07:11:39,0.100000,1.60,960.000000,2.5
5284,2019-03-22 06:24:02,2019-03-22 06:24:11,0.150000,1.70,680.000000,2.5
3241,2019-03-24 17:19:27,2019-03-24 17:19:29,0.033333,0.34,612.000000,65.0


In [16]:
viajes_cortos = taxis[taxis['velocidad_mph'] > 60]
viajes_cortos = viajes_cortos[['duracion_min','distance', 'fare', 'tip', 'velocidad_mph']]
viajes_cortos

,duracion_min,distance,fare,tip,velocidad_mph
400,0.133333,0.90,2.50,0.00,405.000000
491,0.100000,1.60,2.50,0.00,960.000000
770,0.116667,9.40,2.50,0.00,4834.285714
2387,0.550000,1.80,69.06,20.80,196.363636
3241,0.033333,0.34,65.00,13.06,612.000000
3336,0.316667,1.70,14.00,2.00,322.105263
3571,0.150000,0.81,66.00,0.00,324.000000
4746,0.233333,1.30,2.50,0.00,334.285714
5284,0.150000,1.70,2.50,0.00,680.000000
6089,0.300000,10.10,12.00,0.00,2020.000000


### 2.4 Criterio de exclusión (7 pts)

Con base en 1.2, 2.2 y 2.3, define qué viajes excluirías del análisis y por qué. Considera:

a) ¿Todos los viajes de menos de 1 minuto son errores? Observa su distancia y tarifa.

b) ¿Qué umbrales (duración, velocidad, periodo) usarás? Justifícalos.

**Respuesta 2.4:**

_Con base en la actividad 1.2, es necesario descartar el viaje que inicia y termina en febrero, no tiene interés para el análisis de los viajes de Marzo. Por otro lado, yo dejaría el que inicia en marzo y termina en Mayo, se tendría que revisar con los interesados._

_Ahora, con los resultados de las actividades 2.2 y 2.3:_
_No todos los viajes de menos de 1 minuto son errores de registro. Me parece que podemos concluir que los viajes con velocidades superiores a 100 mph y con duración menores a 1 minuto son errores de registro por imposibilidad física de que sucedieran. Adicionalmente, no creo que la información sea concluyente para eliminar más registros, hay viajes muy cortos pero las distancias son coherentes, se podría trabajar un umbral de distancia mínima y máxima para que un viaje sea considerado como viaje._

---
## Actividad 3. Componentes y patrones (25 puntos)

### 3.1 Extraer componentes (6 pts)

Crea las columnas:

- `hora`: hora de `pickup`.
- `dia_num`: día de la semana de `pickup` (lunes = 0).
- `dia_nombre`: nombre del día de `pickup`.
- `fin_semana`: `True` si es sábado o domingo.
- `fecha`: solo la fecha (sin hora) de `pickup`.
- `cruza_medianoche`: `True` si la **fecha** de `dropoff` es distinta de la de `pickup`.

Cuenta cuántos viajes cruzan la medianoche.

In [17]:
# Escribe tu código aquí
taxis['hora'] = taxis['pickup'].dt.hour
taxis['dia_num'] = taxis['pickup'].dt.dayofweek
taxis['dia_nombre'] = taxis['pickup'].dt.day_name()
taxis['fin_semana'] = taxis['dia_num'].isin([5,6])
taxis['fecha'] = taxis['pickup'].dt.date
taxis['cruza_medianoche'] = taxis['fecha'] != taxis['fecha'].shift(1)

taxis.head()

,pickup,dropoff,passengers,distance,fare,tip,tolls,total,color,payment,...,dropoff_borough,duracion,duracion_min,velocidad_mph,hora,dia_num,dia_nombre,fin_semana,fecha,cruza_medianoche
0,2019-03-23 20:21:09,2019-03-23 20:27:24,1,1.60,7.0,2.15,0.0,12.95,yellow,credit card,...,Manhattan,0 days 00:06:15,6.250000,15.360000,20,5,Saturday,True,2019-03-23,True
1,2019-03-04 16:11:55,2019-03-04 16:19:00,1,0.79,5.0,0.00,0.0,9.30,yellow,cash,...,Manhattan,0 days 00:07:05,7.083333,6.691765,16,0,Monday,False,2019-03-04,True
2,2019-03-27 17:53:01,2019-03-27 18:00:25,1,1.37,7.5,2.36,0.0,14.16,yellow,credit card,...,Manhattan,0 days 00:07:24,7.400000,11.108108,17,2,Wednesday,False,2019-03-27,True
3,2019-03-10 01:23:59,2019-03-10 01:49:51,1,7.70,27.0,6.15,0.0,36.95,yellow,credit card,...,Manhattan,0 days 00:25:52,25.866667,17.860825,1,6,Sunday,True,2019-03-10,True
4,2019-03-30 13:27:42,2019-03-30 13:37:14,3,2.16,9.0,1.10,0.0,13.40,yellow,credit card,...,Manhattan,0 days 00:09:32,9.533333,13.594406,13,5,Saturday,True,2019-03-30,True


### 3.2 Patrón por hora del día (7 pts)

Con `groupby('hora')`, calcula para cada hora: el **número de viajes** y la **mediana** de `duracion_min`. Indica la hora con más viajes y la hora con menos.

Usa la mediana (no la media) porque la duración tiene valores extremos.

In [18]:
# Escribe tu código aquí
taxis.groupby('hora').agg({'duracion_min': ['count', 'median']})

duracion_min           
            count     median
hora                        
0             205  10.733333
1             110   9.633333
2             101   9.183333
3              67   9.583333
4              57  10.500000
5              51   8.700000
6             142   7.833333
7             221   9.583333
8             313  11.316667
9             319  11.616667
10            325  11.883333
11            295  12.033333
12            334  12.775000
13            316  10.558333
14            360  11.841667
15            330  12.058333
16            336  11.583333
17            388  11.975000
18            417  11.100000
19            406  10.650000
20            368   9.258333
21            355  11.083333
22            321   9.816667
23            296  10.575000

### 3.3 Patrón por día de la semana (6 pts)

Calcula por día de la semana el **número de viajes** y la **mediana** de `duracion_min` y de `fare`. Presenta los días en orden de lunes a domingo (sugerencia: agrupa por `['dia_num', 'dia_nombre']`).

Toma en cuenta que marzo de 2019 no tiene el mismo número de cada día de la semana: calcula también el **promedio de viajes por fecha**, dividiendo entre el número de fechas distintas de ese día de la semana (usa la columna `fecha` y `.nunique()`).

In [19]:
# Escribe tu código aquí
taxis.groupby(['dia_num', 'dia_nombre']).agg({'duracion_min': ['count', 'median'], 'fare': ['median']})

duracion_min              fare
                          count     median median
dia_num dia_nombre                               
0       Monday              708  10.541667    9.5
1       Tuesday             825  11.216667    9.5
2       Wednesday           966  11.291667    9.5
3       Thursday            905  11.766667    9.5
4       Friday             1115  11.150000    9.5
5       Saturday           1046  10.466667    9.0
6       Sunday              868   9.983333    9.0

### 3.4 Interpretación (6 pts)

a) ¿Qué patrones observas por hora y por día? ¿Coinciden la hora con más viajes y la hora con mayor duración mediana? ¿Qué podría explicarlo?

b) ¿Por qué fue necesario calcular el promedio de viajes por fecha en 3.3 en lugar de comparar solo el total por día de la semana?

**Respuesta 3.4:**

_a) Hay pocas sopresas en los patrones por hora, de madrugada hay más viajes en las horas relacionadas a las fiestas, luego disminuye y durante el día crece poco a poco teniendo su pico de viajes alrededor de las 6:00 pm. La mediana de duración más grande la encontramos al medio día. Por día notamos una tendencia a que los viajes más largos y más caros sean entre semana, teniendo su pico de cantidad de viajes el viernes._

_b) Esto es debido a que no todos hay la misma cantidad de días en todo el mes, por lo que el promedio permite una comparación más representativa._

---
## Actividad 4. Dataset limpio y comparación (20 puntos)

### 4.1 Dataset limpio (6 pts)

Crea `taxis_limpio` aplicando el criterio que definiste en 2.4. Imprime cuántos viajes se excluyeron por cada condición y en total, y verifica que `taxis_limpio` cumple todas las condiciones.

In [20]:
# Escribe tu código aquí
print(f"Número total de viajes antes de la limpieza: {taxis.shape[0]}")

# Condición 1: Viajes con pickup antes de marzo
excluded_feb_mask = taxis['pickup'] < '2019-03-01'
excluded_feb_count = excluded_feb_mask.sum()
print(f"Viajes excluidos con 'pickup' antes de marzo de 2019: {excluded_feb_count}")

# Condición 2: Viajes con velocidad promedio mayor a 100 mph
#Primero, asegurarnos de no incluir NaNs en el conteo de exclusiones. Luego, filtrar por velocidad > 100 mph
excluded_speed_mask = (taxis['velocidad_mph'] > 100) & (taxis['velocidad_mph'].notna())
excluded_speed_count = excluded_speed_mask.sum()
print(f"Viajes excluidos con 'velocidad_mph' > 100 mph: {excluded_speed_count}")

# Combinar las condiciones de exclusión para crear el dataset limpio
# La máscara para incluir viajes será lo contrario de las máscaras de exclusión
include_mask = (~excluded_feb_mask) & (~excluded_speed_mask)
taxis_limpio = taxis[include_mask].copy()

total_excluded_count = taxis.shape[0] - taxis_limpio.shape[0]
print(f"Número total de viajes excluidos: {total_excluded_count}")
print(f"Número total de viajes en 'taxis_limpio': {taxis_limpio.shape[0]}")

# Verificación de que taxis_limpio cumple las condiciones:
# No debería haber ningún viaje con pickup antes de 2019-03-01
print(f"Verificación: Viajes con 'pickup' < 2019-03-01 en taxis_limpio: {(taxis_limpio['pickup'] < '2019-03-01').sum()}")
# No debería haber ningún viaje con velocidad_mph > 100
print(f"Verificación: Viajes con 'velocidad_mph' > 100 en taxis_limpio: {(taxis_limpio['velocidad_mph'] > 100).sum()}")

Número total de viajes antes de la limpieza: 6433
Viajes excluidos con 'pickup' antes de marzo de 2019: 1
Viajes excluidos con 'velocidad_mph' > 100 mph: 11
Número total de viajes excluidos: 12
Número total de viajes en 'taxis_limpio': 6421
Verificación: Viajes con 'pickup' < 2019-03-01 en taxis_limpio: 0
Verificación: Viajes con 'velocidad_mph' > 100 en taxis_limpio: 0


### 4.2 Comparación entre periodos (6 pts)

Con `taxis_limpio`, compara los viajes que inician entre las **7:00 y las 9:59** en **días hábiles** contra los del **mismo horario en fin de semana**. Para cada grupo reporta: número de viajes, mediana de `duracion_min` y mediana de `velocidad_mph`.

In [21]:
# Escribe tu código aquí
# Filtrar viajes que inician entre las 7:00 y 9:59
hora_inicio = 7
hora_fin = 9

periodo_manana_mask = (taxis_limpio['hora'] >= hora_inicio) & (taxis_limpio['hora'] <= hora_fin)
viajes_manana = taxis_limpio[periodo_manana_mask]

# Separar en días hábiles y fines de semana
dias_habiles_manana = viajes_manana[viajes_manana['fin_semana'] == False]
fines_semana_manana = viajes_manana[viajes_manana['fin_semana'] == True]

# Reportar métricas para días hábiles
print("Viajes en días hábiles (7:00-9:59)\n")
print(f"Número de viajes: {dias_habiles_manana.shape[0]}")
print(f"Mediana de duración (min): {dias_habiles_manana['duracion_min'].median():.2f}")
print(f"Mediana de velocidad (mph): {dias_habiles_manana['velocidad_mph'].median():.2f}")

# Reportar métricas para fines de semana
print("\nViajes en fines de semana (7:00-9:59)\n")
print(f"Número de viajes: {fines_semana_manana.shape[0]}")
print(f"Mediana de duración (min): {fines_semana_manana['duracion_min'].median():.2f}")
print(f"Mediana de velocidad (mph): {fines_semana_manana['velocidad_mph'].median():.2f}")

Viajes en días hábiles (7:00-9:59)

Número de viajes: 687
Mediana de duración (min): 11.97
Mediana de velocidad (mph): 9.09

Viajes en fines de semana (7:00-9:59)

Número de viajes: 163
Mediana de duración (min): 7.63
Mediana de velocidad (mph): 13.47


### 4.3 Registro de decisión (8 pts)

Redacta el registro de la limpieza realizada, como lo documentarías en un proyecto real. Incluye:

- **Conversión**: formato utilizado y verificación.
- **Zona horaria**: en qué zona están los datos y cómo lo verificaste.
- **Criterio de exclusión**: condiciones, umbrales y justificación.
- **Resultado**: viajes antes, después y excluidos.
- **Hallazgo de 4.2** y **limitaciones** del análisis.

**Respuesta 4.3:**

### Registro de Decisión de Limpieza de Datos

**1. Conversión de Tipos de Datos:**
*   Las columnas "pickup" y "dropoff" fueron convertidas de tipo "object" a "datetime64[ns]" utilizando "pd.to_datetime()" con el formato explícito '%Y-%m-%d %H:%M:%S'. Esto aseguró una interpretación correcta de las fechas y horas. La verificación posterior (".isna().sum()") confirmó que no se generaron valores "NaT" durante la conversión, lo que indica que todos los registros tenían el formato esperado.

**2. Zona Horaria:**
*   Los datos están registrados en la hora local de Nueva York. Esto se verificó en la Actividad 1.3 al observar que la hora "2:00" del 10 de marzo de 2019 no existía en los registros de "pickup", lo cual es consistente con el cambio al horario de verano en Nueva York, donde el reloj se adelantó de las 02:00 a las 03:00.

**3. Criterio de Exclusión:**
*   Se excluyeron viajes basándose en las siguientes condiciones y justificaciones:
    *   **Viajes fuera de marzo de 2019**: Se excluyó 1 viaje con "pickup" antes del '2019-03-01'. Aunque se identificó 1 viaje con "dropoff" en abril que inició en marzo, se decidió mantenerlo ya que el inicio del viaje sí está dentro del mes de interés.
    *   **Velocidades anómalamente altas**: Se excluyeron los viajes con una "velocidad_mph" superior a 100 mph (identificados 11 viajes). Este umbral se justifica por la imposibilidad física de mantener tales velocidades en un entorno de tráfico urbano como Nueva York, lo que sugiere errores en el registro de tiempo o distancia.

**4. Resultado de la Limpieza:**
*   **Viajes iniciales**: 6433
*   **Viajes excluidos por "pickup" antes de marzo**: 1
*   **Viajes excluidos por "velocidad_mph" > 100**: 11
*   **Total de viajes excluidos**: 12 (incluye viajes que podrían cumplir ambas condiciones o solo una).
*   **Viajes en "taxis_limpio"**: 6421


---
## Reflexión final (10 puntos)

1. ¿Qué ventajas tuvo indicar el formato explícito en 1.1 frente a dejar que pandas lo infiera? ¿En qué situación la inferencia habría producido errores silenciosos?
2. Supón que `pickup` se hubiera registrado en UTC y `dropoff` en hora local de Nueva York. ¿Qué efecto tendría en las duraciones calculadas? ¿Cómo lo habrías detectado con las herramientas de esta actividad?
3. Si se entrenara un modelo para predecir la duración de un viaje, ¿qué problemas causarían las anomalías temporales que encontraste si no se limpiaran?

**Respuesta final:**

_Indicar el formato de fecha explícitamente evitó interpretaciones incorrectas de los datos y posibles errores silenciosos. Si las columnas pickup y dropoff estuvieran en distintas zonas horarias (UTC y hora local), las duraciones calculadas serían erróneas, generando valores anómalos o incluso negativos, algo detectable mediante estadísticas descriptivas y análisis de outliers. Si estas anomalías no se limpiaran antes de entrenar un modelo predictivo, introducirían ruido y sesgos que reducirían la precisión y confiabilidad de las predicciones._

---
**Fin de la actividad.** Antes de entregar: **Reiniciar y ejecutar todo**, verificar que no haya errores, guardar con las salidas visibles y comprobar que el enlace abre sin iniciar sesión.